client/extensions/woocommerce/state/ui/products/edits-reducer.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import { isNumber } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	WOOCOMMERCE_EDIT_PRODUCT,
+} from '../../action-types';
+
+const initialState = null;
+
+export default function( state = initialState, action ) {
+	const handlers = {
+		[ WOOCOMMERCE_EDIT_PRODUCT ]: editProductAction,
+	};
+
+	const handler = handlers[ action.type ];
+
+	return ( handler && handler( state, action ) ) || state;
+}
+
+function editProductAction( edits, action ) {
+	const { product, data } = action.payload;
+
+	const prevEdits = edits || {};
+	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
+	const array = editProduct( prevEdits[ bucket ], product, data );
+
+	return { ...prevEdits, [ bucket ]: array };
+}
+
+function editProduct( array, product, data ) {
+	// Use the existing product id (real or placeholder), or creates.length if no product.
+	const prevArray = array || [];
+	const productId = ( product ? product.id : { index: prevArray.length } );
```

Comment:
In this case for `editProduct` now, an id could be a number if it's an existing product, or it could be an object like `{ index: ## }` if it's a new object that hasn't yet been created on the server. Either way, it's fine. And a product in the edit state will always have an ID from this point forward because it's created here for any new creates (denoted by a null product on edit)